overlord/devicestate/devicestate.go

```diff
@@ -709,6 +709,11 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
 	// go through all the model snaps, see if there are new required snaps
 	// or a track for existing ones needs to be updated
 	for _, modelSnap := range new.SnapsWithoutEssential() {
+		// if the snap isn't required, then we don't do anything right now
+		if modelSnap.Presence != "required" {
```

Comment:
If the snap wasn't required, then `needsInstall` was false. This, in combination with the snap not already being present on the system, resulted in `ts` never being assigned to anything. `ts` was then used on line 770, resulting in a panic inside of that function call.

I think this function could use a refactor to help prevent issues like this. I have a change ready that I will make a PR with. I just wanted to make this bug-fix easier to review.

Reply:
this change doesn't look correct though, if the snap is optional and is present on the device we still want to consider whether it needs a channel change